internal/controller/management/serviceaccount/serviceaccounts.go

```diff
@@ -0,0 +1,289 @@
+	cfg    ReconcilerConfig
+	client client.Client
+
+	getServiceAccountFn func(
+		context.Context,
+		types.NamespacedName,
+		client.Object,
+		...client.GetOption,
+	) error
+
+	createRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.CreateOption,
+	) error
+
+	deleteRoleBindingFn func(
+		context.Context,
+		client.Object,
+		...client.DeleteOption,
+	) error
+
+	ensureControllerPermissionsFn func(context.Context, *corev1.ServiceAccount) error
+	removeControllerPermissionsFn func(context.Context, *corev1.ServiceAccount) error
+}
+
+// SetupReconcilerWithManager initializes a reconciler for ServiceAccount 
+// resources and registers it with the provided Manager.
+func SetupReconcilerWithManager(kargoMgr manager.Manager, cfg ReconcilerConfig) error {
+	return ctrl.NewControllerManagedBy(kargoMgr).
+		For(&corev1.ServiceAccount{}).
+		WithEventFilter(
+			predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					// Allow creation events to be handled for all ServiceAccounts
+					// This ensures any labeled or unlabeled SA gets proper
+					// reconciliation, including on restarts
+					return true
+				},
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					return false
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					oldSA := e.ObjectOld.(*corev1.ServiceAccount)
+					newSA := e.ObjectNew.(*corev1.ServiceAccount)
+					return hasControllerLabel(oldSA) != hasControllerLabel(newSA) || hasControllerLabel(newSA) && newSA.DeletionTimestamp != nil
+				},
+				GenericFunc: func(event.GenericEvent) bool {
+					return false
+				},
+			},
+		).
+		Complete(newReconciler(kargoMgr.GetClient(), cfg))
+}
+
+func newReconciler(kubeClient client.Client, cfg ReconcilerConfig) *reconciler {
+	r := &reconciler{
+		cfg:    cfg,
+		client: kubeClient,
+	}
+	r.getServiceAccountFn = r.client.Get
+	r.createRoleBindingFn = r.client.Create
+	r.deleteRoleBindingFn = r.client.Delete
+	r.ensureControllerPermissionsFn = r.ensureControllerPermissions
+	r.removeControllerPermissionsFn = r.removeControllerPermissions
+	return r
+}
+
+// Reconcile is part of the main Kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+
+	logger := logging.LoggerFromContext(ctx).WithValues(
+		"serviceAccount", req.NamespacedName.Name,
+	)
+	ctx = logging.ContextWithLogger(ctx, logger)
+	logger.Debug("reconciling ServiceAccount")
+
+	sa := &corev1.ServiceAccount{}
+	if err := r.getServiceAccountFn(ctx, req.NamespacedName, sa); err != nil {
+		if kubeerr.IsNotFound(err) {
+			logger.Debug("ServiceAccount not found, deleting RoleBindings for all Projects")
+			// If not found it means the ServiceAccount with the
+			// controller label was deleted, hence remove Controller
+			// Permissions for the ServiceAccount.
+			return ctrl.Result{}, r.removeControllerPermissionsFn(ctx, sa)
```

Comment:
This isn't going to work out the way you hope. If the SA was not found, the `sa` your are passing to `r.removeControllerPermissionsFn()` lacks any details of that SA. It may be that the `removeControllerPermissions()` method needs to be refactored to take a namespaced name as input instead of a SA struct.